overlord/configstate/transaction.go

```diff
@@ -0,0 +1,163 @@
+	if err := state.Get("config", &transaction.data.Config); err != nil {
+		transaction.data.Config = make(systemConfig)
+	}
+
+	return transaction
+}
+
+// Set associates the value with the key for the given snap. Note that this
+// doesn't save any changes (see Commit()), so if the transaction is destroyed
+// the changes have no effect. Also note that the value must properly marshal
+// and unmarshal with encoding/json.
+func (t *Transaction) Set(snapName, key string, value interface{}) error {
+	config, ok := t.data.WriteCache[snapName]
+	if !ok {
+		config = make(snapConfig)
+	}
+
+	// Place the new config value into the snap config
+	marshalledValue, err := json.Marshal(value)
+	if err != nil {
+		return fmt.Errorf("cannot marshal snap %q config value for %q: %s", snapName, key, err)
+	}
+	raw := json.RawMessage(marshalledValue)
+	config[key] = &raw
+
+	// Put that config into the write cache
+	t.data.WriteCache[snapName] = config
+
+	return nil
+}
+
+// Get extracts the stored value associated with the provided key for the snap's
+// config into the value parameter. Note that this query is run against the
+// cached copy of the config; you won't see updates from other transactions.
+func (t *Transaction) Get(snapName, key string, value interface{}) error {
+	// Extract the config for this specific snap. Check the cache first; if it's
+	// not there, check the config from the state
+	if err := t.get(t.data.WriteCache, snapName, key, value); err != nil {
+		return t.get(t.data.Config, snapName, key, value)
+	}
+
+	return nil
+}
+
+// Commit actually saves the changes made to the config in the transaction to
+// the state.
+func (t *Transaction) Commit() {
+	// Do nothing if there's nothing to commit.
+	if len(t.data.WriteCache) == 0 {
+		return
+	}
+
+	t.state.Lock()
+	defer t.state.Unlock()
+
+	// Update our copy of the config with the most recent one from the state.
+	if err := t.state.Get("config", &t.data.Config); err != nil {
+		// Make sure it's still a valid map, in case Get() modified it.
+		t.data.Config = make(systemConfig)
+	}
+
+	// Iterate through the write cache and save each item.
+	for snapName, snapConfigToWrite := range t.data.WriteCache {
+		newSnapConfig, ok := t.data.Config[snapName]
+		if !ok {
+			newSnapConfig = make(snapConfig)
+		}
+
+		for key, value := range snapConfigToWrite {
+			newSnapConfig[key] = value
+		}
+
+		t.data.Config[snapName] = newSnapConfig
+	}
+
+	t.state.Set("config", t.data.Config)
+
+	// The cache has been flushed-- reset it
+	t.data.WriteCache = make(systemConfig)
+}
+
+func (t *Transaction) MarshalJSON() ([]byte, error) {
+	return json.Marshal(t.data)
+}
+
+func (t *Transaction) UnmarshalJSON(bytes []byte) error {
+	return json.Unmarshal(bytes, &t.data)
+}
+
+func (t *Transaction) get(config systemConfig, snapName, key string, value interface{}) error {
+	c, ok := config[snapName]
+	if !ok {
+		return fmt.Errorf("no config available for snap %q", snapName)
```

Comment:
This and the case below sound like the same error from a user perspective. I suggest something like:
```
"snap %q has no %q configuration option"
```
for both.